### PR TITLE
zip_extract_stream()

### LINKED ISSUE #126: Expose a function that opens an archive that is already loaded in memory.
```zip_open``` works only for archives that come from the file system.

### PR DESCRIPTION
**Pull request by jinfeihan57**
Fixes #126  . Part 1
zip_open_stream() I will do it later. 
@kuba-- 
But first Check whether zip_extract_stream() is appropriate.

### CODE REVIEW

File: src/zip.c [hunk 1]
```diff
@@ -957,3 +957,129 @@ int zip_extract(const char *zipname, const char *dir,
 
   return status;
 }
+
+int zip_extract_stream(const char *zipstream, size_t size, const char *dir,
```
Comment (kuba--) [inline src/zip.c:961]: Could you rename this arg, to just `stream`

File: src/zip.c [hunk 2]
```diff
@@ -957,3 +957,129 @@ int zip_extract(const char *zipname, const char *dir,
...
+
+  memset(path, 0, sizeof(path));
+  memset(symlink_to, 0, sizeof(symlink_to));
+  if (!memset(&(zip_archive), 0, sizeof(zip_archive))) {
+    // Cannot memset zip archive
+    return -1;
+  }
+
+  if (!zipstream || !dir) {
+    // Cannot parse zip archive name
```
Comment (kuba--) [inline src/zip.c:980]: This comment is incorrect here here.

Comment (jinfeihan57) [reply src/zip.c:980]: Yes，I'll fix it.

File: src/zip.c (hunk 1 above)
Comment (jinfeihan57) [reply src/zip.c:961]: what do you mean? 
 Variable： void * arg ?？

**Comment by jinfeihan57** [comment]
> Generally LGTM, but because the main difference between `zip_extract` and `zip_extract_stream` is the line:
> 
> ```c
> if (!mz_zip_reader_init_mem(&zip_archive, zipstream, size, 0)) {
> ```
> 
> So, I suggest to to put the common code into some static function and call it in both function, instead of copy-paste.
Yes，I'll try to change the duplicate code to a static function.

File: src/zip.c (hunk 1 above)
Comment (jinfeihan57) [reply src/zip.c:961]: Do you mean zipstream -> stream?

Comment (jinfeihan57) [reply src/zip.c:961]: In line 602 : FILE *stream = NULL;
arg stream means FILE *. 
so I named it zipstream.

Comment (kuba--) [reply src/zip.c:961]: yeah, but it's a different function and local var. so we can rename to something else.

File: test/test.c [hunk 3]
```diff
@@ -462,6 +465,35 @@ static void test_unix_permissions(void) {
...
+
+  FILE *fp = NULL;
+  fp = fopen(ZIPNAME, "r+");
+  assert(fp != NULL);
+ 
+  fseek(fp, 0L, SEEK_END);
+  size_t filesize = ftell(fp);
+  fseek(fp, 0L, SEEK_SET);
+
+  char zipstream[filesize];
```
[the remark this answers is not in the document]
Comment (kuba--) [reply test/test.c:488]: We don't test _big files_ here, but just accuracy of the test.
So, actually  here (in tests, where we don't want to worry about dealloc memory) we can define  `MAXBUFSIZE` macro big enough and just set `char zipstream[MAXBUFSIZE];`, but remember to `memset` it first.

File: test/test.c [hunk 4]
```diff
@@ -20,6 +20,9 @@
 #define TESTDATA2 "Some test data 2...\0"
 #define CRC32DATA2 2532008468
 
+#define DIREMPTY "empty\0"
```
[the remark this answers is not in the document]
Comment (jinfeihan57) [reply test/test.c:23]: Sorry，I forget to delete it.

File: test/test.c [hunk 5]
```diff
@@ -20,6 +20,9 @@
...
+#define DIRTEST "test\0"
```
[the remark this answers is not in the document]
Comment (jinfeihan57) [reply test/test.c:24]: Sorry，I forget to delete it.

File: src/zip.c (hunk 1 above)
Comment (jinfeihan57) [reply src/zip.c:961]: Ok.

File: test/test.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (jinfeihan57) [reply test/test.c:488]: I'll memset() it to 0 before using it.

File: test/test.c [hunk 6]
```diff
@@ -462,6 +465,35 @@ static void test_unix_permissions(void) {
...
+  assert(0 == zip_create(ZIPNAME, filenames, 1));
+  remove(XFILE);
+
+  FILE *fp = NULL;
+  fp = fopen(ZIPNAME, "r+");
+  assert(fp != NULL);
+ 
+  fseek(fp, 0L, SEEK_END);
+  size_t filesize = ftell(fp);
+  fseek(fp, 0L, SEEK_SET);
```
[the remark this answers is not in the document]
Comment (jinfeihan57) [reply test/test.c:486]: zip_extract_stream() only extract archive to file system not memory.
zip_open_stream() can exrtact arcive to memory. than we can compare to the original buffer.

**Comment by jinfeihan57** [comment]
@kuba-- 
Please check again.

File: src/zip.c [hunk 7]
```diff
@@ -950,10 +934,88 @@ int zip_extract(const char *zipname, const char *dir,
...
+                                      size_t size, mz_uint32 flags){
+  mz_zip_archive *zip_archive = NULL;
+  zip_archive = (mz_zip_archive *)malloc(sizeof(mz_zip_archive));
+  if (!zip_archive){
+    // malloc failed.
+    return NULL;
+  }
+  if (!memset(zip_archive, 0, sizeof(mz_zip_archive))) {
+    // Cannot memset zip archive
+    return NULL;
```
Comment (kuba--) [inline src/zip.c:993]: See comments above regarding `malloc` and `free` if we fail  after successful `malloc`

File: src/zip.h [hunk 8]
```diff
@@ -313,6 +313,25 @@ extern int zip_extract(const char *zipname, const char *dir,
...
+ *
+ * @param zipstream zip archive stream.
+ * @param size stream size.
+ * @param dir output directory.
+ * @param on_extract on extract callback.
+ * @param arg opaque pointer.
+ *
+ * @return the return code - 0 on success, negative number (< 0) on error.
+ */
+extern int zip_extract_stream(const char *zipstream, size_t size, const char *dir,
```
Comment (kuba--) [inline src/zip.h:333]: how about renaming `zipstream` just to `stream`?

File: test/test.c [hunk 9]
```diff
@@ -462,6 +462,41 @@ static void test_unix_permissions(void) {
...
+
+  char stream[filesize];
+  memset(stream, 0, filesize);
+  size_t size = fread(stream, 1, filesize, fp);
```
Comment (kuba--) [inline test/test.c:492]: just in case replace `1` by `sizeof(char)`

File: src/zip.c [hunk 10]
```diff
@@ -950,10 +934,88 @@ int zip_extract(const char *zipname, const char *dir,
...
+static mz_zip_archive * zip_archive_init_file_reader(const char *zipname, mz_uint32 flags){
+  mz_zip_archive *zip_archive = NULL;
+  zip_archive = (mz_zip_archive *)malloc(sizeof(mz_zip_archive));
+  if (!zip_archive){
+    // malloc failed.
+    return NULL;
+  }
+  if (!memset(zip_archive, 0, sizeof(mz_zip_archive))) {
+    // Cannot memset zip archive
+    return NULL;
```
[the remark this answers is not in the document]
Comment (codylico) [reply src/zip.c:954]: (outsider comment, feel free to ignore) Can `memset` fail here? According to [cppreference.com](https://en.cppreference.com/w/c/string/byte/memset), `memset` just returns the pointer passed in. It doesn't really have a failure case (except for access violation, etc.).

[the remark this answers is not in the document]
Comment (kuba--) [reply src/zip.c:954]: yeah, memset rather _should not_ fail.

File: src/zip.c [hunk 11]
```diff
@@ -832,40 +832,24 @@ int zip_create(const char *zipname, const char *filenames[], size_t len) {
   return status;
 }
 
-int zip_extract(const char *zipname, const char *dir,
-                int (*on_extract)(const char *filename, void *arg), void *arg) {
+static int zip_extract_internal(mz_zip_archive *zip_archive, const char *dir,
```
[the remark this answers is not in the document]
Comment (jinfeihan57) [reply src/zip.c:835]: We already have an API called zip_extract, so I'm going to use zip_extract_unzip.

File: src/zip.c [hunk 12]
```diff
@@ -950,10 +934,88 @@ int zip_extract(const char *zipname, const char *dir,
...
-  if (!mz_zip_reader_end(&zip_archive)) {
+  if (!mz_zip_reader_end(zip_archive)) {
     // Cannot end zip reader
     status = -1;
   }
+  return status;
+
+}
+
+static mz_zip_archive * zip_archive_init_file_reader(const char *zipname, mz_uint32 flags){
```
[the remark this answers is not in the document]
Comment (jinfeihan57) [reply src/zip.c:945]: ok.

File: src/zip.h (hunk 8 above)
Comment (jinfeihan57) [reply src/zip.h:333]: ok.

File: test/test.c (hunk 9 above)
Comment (jinfeihan57) [reply test/test.c:492]: ok.

File: src/zip.c (hunk 7 above)
Comment (jinfeihan57) [reply src/zip.c:993]: ok.

File: src/zip.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (jinfeihan57) [reply src/zip.c:954]: ok.

File: src/zip.c (hunk 12 above)
[the remark this answers is not in the document]
Comment (jinfeihan57) [reply src/zip.c:945]: One more question.
zip_archive_init_file or just init_file.case I think "zip_archive" prefix is needed.

**Comment by jinfeihan57** [comment]
@kuba--

**Comment by kuba--** [comment]
@jinfeihan57 - I reverted your PR, because tests don't compile on windows, see CI warnings:
```
D:\a\zip\zip\test\test.c(490,23): error C2057: expected constant expression [D:\a\zip\zip\test\test.out.vcxproj]
 [27](https://github.com/kuba--/zip/runs/997826701#step:4:27) 
D:\a\zip\zip\test\test.c(490,23): error C2466: cannot allocate an array of constant size 0 [D:\a\zip\zip\test\test.out.vcxproj]
 [28](https://github.com/kuba--/zip/runs/997826701#step:4:28) 
D:\a\zip\zip\test\test.c(490,24): error C2133: 'stream': unknown size [D:\a\zip\zip\test\test.out.vcxproj]
 [29](https://github.com/kuba--/zip/runs/997826701#step:4:29) 
```

Also remember - before you PR, to reformat your code:
```
for file in $(git ls-files | \grep -E '\.(c|h)$' | \grep -v -- '#')
do
    clang-format -i $file
done
```

So please fix those issues and send a new PR.

btw. MacOS builds are failing because github changed some linker configuration (I'll have to dig in)

**Pull request merged**